components/canvas/webgl_paint_task.rs

```diff
@@ -0,0 +1,155 @@
+                    },
+                    CanvasMsg::ClearColor(r, g, b, a) => {
+                      painter.clear_color(r, g, b, a);
+                    },
+                    CanvasMsg::Close => break,
+                    CanvasMsg::Recreate(size) => painter.recreate(size),
+                    CanvasMsg::Render => {
+                      painter.render();
+                    },
+                    CanvasMsg::SendPixelContents(chan) => painter.send_pixel_contents(chan),
+                    _ => panic!("Wrong message sent to WebGLTask"),
+                }
+            }
+        });
+        chan
+    }
+
+    fn init(&self) {
+        let mut framebuffer_ids = vec!();
+        let mut texture_ids = vec!();
+        framebuffer_ids = gl::gen_framebuffers(1);
+        gl::bind_framebuffer(gl::FRAMEBUFFER, framebuffer_ids[0]);
+
+        texture_ids = gl::gen_textures(1);
+        gl::bind_texture(gl::TEXTURE_2D, texture_ids[0]);
+
+        gl::tex_image_2d(gl::TEXTURE_2D, 0, gl::RGB as GLint, self.size.width as GLsizei,
+                         self.size.height as GLsizei, 0, gl::RGB, gl::UNSIGNED_BYTE, None);
+        gl::tex_parameter_i(gl::TEXTURE_2D, gl::TEXTURE_MAG_FILTER, gl::NEAREST as GLint);
+        gl::tex_parameter_i(gl::TEXTURE_2D, gl::TEXTURE_MIN_FILTER, gl::NEAREST as GLint);
+
+        gl::framebuffer_texture_2d(gl::FRAMEBUFFER, gl::COLOR_ATTACHMENT0, gl::TEXTURE_2D,
+                                   texture_ids[0], 0);
+        gl::bind_texture(gl::TEXTURE_2D, 0);
+
+        gl::viewport(0 as GLint, 0 as GLint,
+                     self.size.width as GLsizei, self.size.height as GLsizei);
+    }
+
+    fn clear(&self, mask: u32) {
+        gl::clear(mask);
+        self.render();
+    }
+
+    fn clear_color(&self, r: f32, g: f32, b: f32, a: f32) {
+        gl::clear_color(r, g, b, a);
+    }
+
+    fn send_pixel_contents(&mut self, chan: Sender<Vec<u8>>) {
+        self.drawtarget.snapshot().get_data_surface().with_data(|element| {
+            chan.send(element.to_vec()).unwrap();
+        })
+    }
+
+    fn recreate(&mut self, size: Size2D<i32>) {
+        self.drawtarget = WebGLPaintTask::create(size,
+                                                 &self.native_graphics_context);
+    }
+
+    fn render(&self) {
+        let mut pixels = gl::read_pixels(0, 0,
+                                    self.size.width as gl::GLsizei,
+                                    self.size.height as gl::GLsizei,
+                                    gl::RGBA, gl::UNSIGNED_BYTE);
+
+        // rgba -> bgra
+        byte_swap(pixels.as_mut_slice());
+
+        let source_surface = self.drawtarget.create_source_surface_from_data(&pixels,
+            self.size, self.size.width * 4, SurfaceFormat::B8G8R8A8);
+
+        let canvas_rect = Rect(Point2D(0i32, 0i32), self.size);
+        let draw_surface_options = DrawSurfaceOptions::new(Filter::Linear, true);
+        let draw_options = DrawOptions::new(1.0f64 as AzFloat, 0);
```

Comment:
nit: `1.0` is sufficient (but I don't think you need this code at all, see above).